File: Algorithms_easy/112._Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
            if not root:
                return False
            targetSum -= root.val
            if not root.left and not root.right:
                return targetSum == 0
            return self.hasPathSum(root.left, targetSum) or self.hasPathSum(root.right, targetSum)
'''
直接开始遍历，targetSum减去遇到的数，看是否有一条枝干符合条件
递归地询问它的子节点是否能满足条件即可
'''

File: Algorithms_C++/112._Path_Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return dfs(root, targetSum, 0);
    }
    bool dfs(TreeNode* root, int targetSum, int s){
        if(!root){
            return false;
        }
        s += root->val;
        if(!root->left && !root->right && s == targetSum){
            return true;
        }
        return dfs(root->left, targetSum, s) || dfs(root->right, targetSum, s);
    }
};